adapters/play/src/test/scala/caliban/AdapterSpec.scala

```diff
@@ -0,0 +1,214 @@
+                         } yield fileUploadResp
+                       }
+                     )
+        } yield response.body
+
+        assertM(body.map(_.toOption.get.data.uploadFile))(
+          hasField("hash", (f: TestAPI.File) => f.hash, equalTo(fileHash)) &&
+            hasField("filename", (f: TestAPI.File) => f.filename, equalTo(fileName)) &&
+            hasField("mimetype", (f: TestAPI.File) => f.mimetype, equalTo("image/png"))
+        )
+      },
+      testM("multipart request with several files") {
+        val file1Hash         = "64498927ff9cd735daefebe7175ed1567650399e58648a6b8340f636243962c0"
+        val file1Name: String = s"$file1Hash.png"
+        val file1URL: URL     = getClass.getResource(s"/$file1Name")
+
+        val file2Hash         = "d6359a52607b6953b2cb96be00b228c6030f11806e380c29c3f5d8db608c399b"
+        val file2Name: String = s"$file2Hash.txt"
+        val file2URL: URL     = getClass.getResource(s"/$file2Name")
+
+        val query: String =
+          """{ "query": "mutation ($files: [Upload!]!) { uploadFiles(files: $files) { hash, path, filename, mimetype } }",   "variables": {  "files": [null, null] }}"""
+
+        val request = basicRequest
+          .post(uri)
+          .contentType("multipart/form-data")
+          .multipartBody(
+            multipart("operations", query).contentType("application/json"),
+            multipart("map", """{ "0": ["variables.files.0"], "1":  ["variables.files.1"]}"""),
+            multipartFile("0", new File(file1URL.getPath)).contentType("image/png"),
+            multipartFile("1", new File(file2URL.getPath)).contentType("text/plain")
+          )
+
+        val body = for {
+          response <- SttpClient.send(
+                       request.mapResponse { strRespOrError =>
+                         for {
+                           resp           <- strRespOrError
+                           json           <- parse(resp)
+                           fileUploadResp <- json.as[Response[UploadFiles]]
+                         } yield fileUploadResp
+                       }
+                     )
+        } yield response.body
+
+        assertM(body.map(_.toOption.get.data.uploadFiles))(
+          hasField("hash", (fl: List[TestAPI.File]) => fl(0).hash, equalTo(file1Hash)) &&
+            hasField("hash", (fl: List[TestAPI.File]) => fl(1).hash, equalTo(file2Hash)) &&
+            hasField("filename", (fl: List[TestAPI.File]) => fl(0).filename, equalTo(file1Name)) &&
+            hasField("filename", (fl: List[TestAPI.File]) => fl(1).filename, equalTo(file2Name)) &&
+            hasField("mimetype", (fl: List[TestAPI.File]) => fl(0).mimetype, equalTo("image/png")) &&
+            hasField("mimetype", (fl: List[TestAPI.File]) => fl(1).mimetype, equalTo("text/plain"))
+        )
+      }
+    ).provideCustomLayer(AsyncHttpClientZioBackend.layer())
+      .provideCustomLayerShared(specLayer)
```

Comment:
How does this work, wouldn't the first `provideCustomLayer` erase the requirements for everything _except_ `ZEnv` but then you provide the additional `specLayer` is it because the tests don't actually require it? We are just running for the effect?

Reply:
is this not how actually the layer is "executed"?

Reply:
I refactored slightly, but I still don't understand what you mean by erase

Reply:
I meant because the signature for `provideCustomLayer` is:

```scala
  final def provideCustomLayer[E1 >: E, R1 <: Has[_]](
    layer: ZLayer[ZEnv, E1, R1]
  )(implicit ev: ZEnv with R1 <:< R, tagged: Tag[R1]): ZIO[ZEnv, E1, A]
```

So the resulting effect doesn't _require_ the stuff from `specLayer` (at least according to the signature), so I was just curious if it was because we are just running the `Layer` for an effect rather than to get the resource it creates.

Reply:
I guess you are right then. In this case we run for the effect.